data-prepper-plugins/otel-trace-raw-prepper/src/main/java/com/amazon/dataprepper/plugins/prepper/oteltrace/OTelTraceRawPrepper.java

```diff
@@ -94,11 +120,122 @@ public OTelTraceRawPrepper(final PluginSetting pluginSetting) {
+        rawSpans.addAll(getTracesToFlushByGarbageCollection());
+
+        return convertRawSpansToJsonRecords(rawSpans);
+    }
+
+    private List<Record<String>> convertRawSpansToJsonRecords(final List<RawSpan> rawSpans) {
+        final List<Record<String>> records = new LinkedList<>();
+
+        for (RawSpan rawSpan : rawSpans) {
+            String rawSpanJson;
+            try {
+                rawSpanJson = rawSpan.toJson();
+            } catch (JsonProcessingException e) {
+                log.error("Unable to process invalid Span {}:", rawSpan, e);
+                spanErrorsCounter.increment();
+                totalProcessingErrorsCounter.increment();
+                continue;
+            }
+
+            records.add(new Record<>(rawSpanJson));
+        }
+
+        return records;
+    }
+
+    private List<RawSpan> getTracesToFlushByParentSpan() {
+        final List<RawSpan> recordsToFlush = new LinkedList<>();
+
+        for (DelayedParentSpan delayedParentSpan; (delayedParentSpan = delayedParentSpanQueue.poll()) != null;) {
+            RawSpan parentSpan = delayedParentSpan.getRawSpan();
+            recordsToFlush.add(parentSpan);
+
+            String traceGroup = parentSpan.getTraceGroup();
+            String parentSpanTraceId = parentSpan.getTraceId();
+
+            RawSpanSet rawSpanSet = traceIdRawSpanSetMap.get(parentSpanTraceId);
+            if (rawSpanSet != null) {
+                for (RawSpan rawSpan : rawSpanSet.getRawSpans()) {
+                    rawSpan.setTraceGroup(traceGroup);
+                    recordsToFlush.add(rawSpan);
+                }
+
+                traceIdRawSpanSetMap.remove(parentSpanTraceId);
+            }
+        }
+
+        return recordsToFlush;
+    }
+
+
+    private List<RawSpan> getTracesToFlushByGarbageCollection() {
+        final List<RawSpan> recordsToFlush = new LinkedList<>();
+
+        if (shouldGarbageCollect()) {
+            final long now = System.currentTimeMillis();
+            lastGarbageCollectionTime = now;
+
+            // fail-safe
+            Iterator<Map.Entry<String, RawSpanSet>> entryIterator = traceIdRawSpanSetMap.entrySet().iterator();
+            while (entryIterator.hasNext()) {
+                Map.Entry<String, RawSpanSet> entry = entryIterator.next();
+                RawSpanSet rawSpanSet = entry.getValue();
+                long traceTime = rawSpanSet.getTimeSeen();
+                if (now - traceTime >= gcInterval) {
+                    Set<RawSpan> rawSpans = rawSpanSet.getRawSpans();
+                    for (RawSpan rawSpan : rawSpans) {
+                        rawSpan.setTraceGroup("ERROR");
+                        recordsToFlush.add(rawSpan);
+                    }
+
+                    entryIterator.remove();
+                }
+            }
+            if (recordsToFlush.size() > 0) {
+                log.error("Flushing {} records due to GC", recordsToFlush.size());
```

Comment:
This can just be info. Originally I had it as error while testing because I was having trouble with logging thresholds (info wouldn't show up for me, it should be fixed now though).